core/src/main/java/ai/timefold/solver/core/impl/solver/termination/AdaptiveTermination.java

```diff
@@ -1,57 +1,71 @@
         this.minimumImprovementRatio = minimumImprovementRatio;
-        this.scoresByTime = new TreeMap<>();
-    }
-
-    private record LevelScoreDiff(boolean harderScoreChanged, double softestScoreDiff) {
-        /**
-         * Calculates the softest score difference between two scores and records
-         * if any harder levels changed. Returns null if the score are the same.
-         *
-         * @param start The first score, typically smaller
-         * @param end The second score, typically larger
-         * @return A {@link LevelScoreDiff} where harderScoreChanged is true
-         *         if and only if a level beside the softest score level changed,
-         *         and the difference between the softest score as a double, or null
-         *         if both scores are the same.
-         * @param <Score_> The score type.
-         */
-        public static <Score_ extends Score<Score_>> @Nullable LevelScoreDiff between(@NonNull Score_ start,
-                @NonNull Score_ end) {
-            var scoreDiffs = end.subtract(start).toLevelDoubles();
-            var softestLevel = scoreDiffs.length - 1;
-            for (int i = 0; i < scoreDiffs.length; i++) {
-                if (scoreDiffs[i] != 0.0) {
-                    return new LevelScoreDiff(softestLevel != i, scoreDiffs[softestLevel]);
-                }
+        this.scoresByTime = new AdaptiveScoreRingBuffer<>();
+    }
+
+    /**
+     * An exception that is thrown to signal a hard score improvement.
+     * This is done since:
+     * <p/>
+     * <ul>
+     * <li/>Hard score improvements are significantly rarer than soft score improvements.
+     * <li/>Allows us to directly return a double in a method instead of using a
+     * carrier type/needing to box it so we can use null as a special value.
+     * <li/>Avoid branching in code on the hot path.
+     * </ul>
+     * <p/>
+     * This exception does not fill the stack trace to improve performance;
+     * it is a signal that should always be caught
+     * and handled appropriately.
+     */
+    private static final class HardLevelImprovedSignal extends Exception {
+        @Override
+        public Throwable fillInStackTrace() {
+            return this;
+        }
+    }
+
+    private static <Score_ extends Score<Score_>> double softImprovement(@NonNull Score_ start,
+            @NonNull Score_ end) throws HardLevelImprovedSignal {
+        if (start.equals(end)) {
+            // optimization: since most of the time the score the same,
+            // we can use equals to avoid creating double arrays in the
+            // vast majority of comparisons
+            return 0.0;
+        }
+        var scoreDiffs = end.subtract(start).toLevelDoubles();
+        var softestLevel = scoreDiffs.length - 1;
+        for (int i = 0; i < softestLevel; i++) {
+            if (scoreDiffs[i] != 0.0) {
+                throw new HardLevelImprovedSignal();
```

Comment:
May I suggest we return an impossible value instead of using exceptions for control flow? Perhaps [NaN](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Double.html#NaN)?

Reply:
I understand your reasoning, but I disagree. In my opinion, what you're doing here is misusing exceptions - you want to terminate without a return, and exceptions are a _nice_ way of doing it. When you have a hammer, everything becomes a nail.

Exceptions are meant for _exceptional_ cases - this situation is not exceptional.
Exceptions are meant for reporting errors, not for control flow. This is not an error.
And finally, unless I'm mistaken, exceptions mess with JVM's optimizations, because they abort the flow.

I could go on, but I'd just be duplicating what was already said much better in [the Java bible](https://www.amazon.com/Effective-Java-Joshua-Bloch/dp/0134685997). :-)

Reply:
Good, data is always better than no data. Doesn't this in fact prove that using the exception is the slower option, as I expected?

> that there is no way to enforce the caller to actually check it is NaN.

The argument would have value if this was public API where there was a risk that the user can misuse it. However, this is a private method inside of an internal implementation - external users don't enter the picture, and for our purposes we make sure it works through tests.

Reply:
> Good, data is always better than no data. Doesn't this in fact prove that using the exception is the slower option, as I expected?

Yes (although I wouldn't use prove; this is only a single benchmark run, on a single machine).

> > that there is no way to enforce the caller to actually check it is NaN.
> 
> The argument would have value if this was public API where there was a risk that the user can misuse it. However, this is a private method inside of an internal implementation - external users don't enter the picture, and for our own purposes we make sure it works through tests.

I will rename the method to something like `softImprovementOrNaNOnHardImprovement` to prevent headaches for future me.